### PR TITLE
BUG: Use size_t to avoid array index overflow; add missing malloc of error_msg

### LINKED ISSUE #16798: Pandas core dumps when reading large CSV file using read_csv(..., low_memory=False)
#### Code Sample, a copy-pastable example if possible

```python
    try:
        table = pd.read_csv(csv_file_name, low_memory=False)
    except:
        raise

```
#### Problem description

From the stacktrace in the core file, pandas seems to be throwing an exception complaining "out of memory" (which it is not, the machine has 64 G of RAM and the interpreter was using maybe 5 G) but, during the cleanup of that exception, attempts to double free the `self->error_msg` pointer (according to gcc). Results in a `SIGSEGV`.

#### Expected Output

Pandas successfully converts the CSV into a dataframe

#### Output of ``pd.show_versions()``

<details>
>>> pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-81-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.19.2
nose: None
pip: 9.0.1
setuptools: 36.0.1
Cython: 0.25.2
numpy: 1.13.0
scipy: None
statsmodels: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.6.0
pytz: 2016.10
blosc: None
bottleneck: None
tables: None
numexpr: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.1.4
pymysql: None
psycopg2: 2.6.1 (dt dec pq3 ext lo64)
jinja2: 2.9.6
boto: 2.47.0
pandas_datareader: None

</details>

I can provide the source CSV if necessary (though it happens reliably with "large" CSVs, for a definition of "large" I haven't nailed down but in the multi-GB range). Below is the stack trace:

<details>
#0  0x00007f278e970532 in __GI___libc_free (mem=0x7f2757218b0e) at malloc.c:2967
#1  0x00007f275720fd2a in free_if_not_null (ptr=0x3cf5ee8) at pandas/src/parser/tokenizer.c:94
#2  parser_cleanup (self=self@entry=0x3cf5df0) at pandas/src/parser/tokenizer.c:189
#3  0x00007f275720ff09 in parser_free (self=0x3cf5df0) at pandas/src/parser/tokenizer.c:285
#4  0x00007f27571b0562 in __pyx_pf_6pandas_6parser_10TextReader_4__dealloc__ (__pyx_v_self=0x7f2757166bc8) at pandas/parser.c:6330
#5  __pyx_pw_6pandas_6parser_10TextReader_5__dealloc__ (__pyx_v_self=<pandas.parser.TextReader at remote 0x7f2757166bc8>) at pandas/parser.c:6313
#6  __pyx_tp_dealloc_6pandas_6parser_TextReader (o=<pandas.parser.TextReader at remote 0x7f2757166bc8>) at pandas/parser.c:45130
#7  0x000000000055dbea in dict_dealloc.lto_priv.164 (mp=0x7f2753a0bbc8) at ../Objects/dictobject.c:1594
#8  subtype_dealloc () at ../Objects/typeobject.c:1193
#9  0x000000000055dbea in dict_dealloc.lto_priv.164 (mp=0x7f2753c765c8) at ../Objects/dictobject.c:1594
#10 subtype_dealloc () at ../Objects/typeobject.c:1193
#11 0x00000000004e9137 in frame_dealloc.lto_priv () at ../Objects/frameobject.c:431
#12 0x0000000000541457 in tb_dealloc.lto_priv.286 (tb=0x7f27537b8048) at ../Python/traceback.c:55
#13 0x000000000054146d in tb_dealloc.lto_priv.286 (tb=0x7f27537b8088) at ../Python/traceback.c:54
#14 0x000000000054146d in tb_dealloc.lto_priv.286 (tb=0x7f27537b80c8) at ../Python/traceback.c:54
#15 0x000000000054146d in tb_dealloc.lto_priv.286 (tb=0x7f27537b8108) at ../Python/traceback.c:54
#16 0x0000000000526b99 in PyEval_EvalFrameEx () at ../Python/ceval.c:2132
#17 0x0000000000528814 in fast_function (nk=<optimized out>, na=<optimized out>, n=<optimized out>, pp_stack=0x7ffdc3b87590, func=<optimized out>) at ../Python/ceval.c:4803
#18 call_function (oparg=<optimized out>, pp_stack=0x7ffdc3b87590) at ../Python/ceval.c:4730
#19 PyEval_EvalFrameEx () at ../Python/ceval.c:3236
#20 0x0000000000528814 in fast_function (nk=<optimized out>, na=<optimized out>, n=<optimized out>, pp_stack=0x7ffdc3b876c0, func=<optimized out>) at ../Python/ceval.c:4803
#21 call_function (oparg=<optimized out>, pp_stack=0x7ffdc3b876c0) at ../Python/ceval.c:4730
#22 PyEval_EvalFrameEx () at ../Python/ceval.c:3236
#23 0x000000000052d2e3 in _PyEval_EvalCodeWithName () at ../Python/ceval.c:4018
#24 0x000000000052dfdf in PyEval_EvalCodeEx () at ../Python/ceval.c:4039
#25 PyEval_EvalCode (co=<optimized out>, globals=<optimized out>, locals=<optimized out>) at ../Python/ceval.c:777
#26 0x00000000005fd2c2 in run_mod () at ../Python/pythonrun.c:976
#27 0x00000000005ff76a in PyRun_FileExFlags () at ../Python/pythonrun.c:929
#28 0x00000000005ff95c in PyRun_SimpleFileExFlags () at ../Python/pythonrun.c:396
#29 0x000000000063e7d6 in run_file (p_cf=0x7ffdc3b87930, filename=0x1e96260 L"/home/jknupp/lion/components/api/scripts/parquet_export.py", fp=0x1fa1310) at ../Modules/main.c:318
#30 Py_Main () at ../Modules/main.c:768
#31 0x00000000004cfe41 in main () at ../Programs/python.c:65
#32 0x00007f278e90c830 in __libc_start_main (main=0x4cfd60 <main>, argc=2, argv=0x7ffdc3b87b48, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7ffdc3b87b38)
    at ../csu/libc-start.c:291
#33 0x00000000005d5f29 in _start ()
</details>

### PR DESCRIPTION
**Pull request by jeffknupp**
Fix a few locations where a parser's `error_msg` buffer is written to
without having been previously allocated. This manifested as a double
free during exception handling code making use of the `error_msg`.

Additionally, use `size_t/ssize_t` where array indices or lengths will
be stored. Previously, int32_t was used and would overflow on columns
with very large amounts of data (i.e. greater than INTMAX bytes).

 - [ ] closes #16798 
 - [ ] tests added / passed
 - [ ] passes ``git diff upstream/master -u -- "*.py" | flake8 --diff``
 - [ ] whatsnew entry

### CODE REVIEW

File: pandas/_libs/parsers.pyx [hunk 1]
```diff
@@ -121,30 +121,30 @@ cdef extern from "parser/tokenizer.h":
         io_callback cb_io
         io_cleanup cb_cleanup
 
-        int chunksize  # Number of bytes to prepare for each chunk
+        size_t chunksize  # Number of bytes to prepare for each chunk
         char *data     # pointer to data to be processed
-        int datalen    # amount of data available
-        int datapos
+        size_t datalen    # amount of data available
```
Comment (gfyoung) [inline pandas/_libs/parsers.pyx]: Fix the spacing so that the hashtags line up like before.

File: pandas/_libs/parsers.pyx [hunk 2]
```diff
@@ -121,30 +121,30 @@ cdef extern from "parser/tokenizer.h":
...
         # Store words in (potentially ragged) matrix for now, hmm
         char **words
-        int *word_starts # where we are in the stream
-        int words_len
-        int words_cap
+        size_t *word_starts # where we are in the stream
+        size_t words_len
+        size_t words_cap
 
         char *pword_start    # pointer to stream start of current field
```
Comment (gfyoung) [inline pandas/_libs/parsers.pyx]: Modify the spacing so that it's realigned with the hashtags below.

File: pandas/_libs/parsers.pyx [hunk 3]
```diff
@@ -178,13 +178,13 @@ cdef extern from "parser/tokenizer.h":
         char thousands
 
         int header # Boolean: 1: has header, 0: no header
-        int header_start # header row start
-        int header_end # header row end
+        ssize_t header_start # header row start
+        ssize_t header_end # header row end
```
Comment (gfyoung) [inline pandas/_libs/parsers.pyx]: Modify the spacing so that it aligns with the hashtag above.

Comment (gfyoung) [inline pandas/_libs/parsers.pyx]: Also, why `ssize_t` here (and not `size_t` like above)?  We should either use one or the other IMO with these changes.  I think either would work since technically, all of these fields must be >= 0.

**Comment by wesm** [comment]
Very honestly, I'm not sure I'm comfortable with this many unsigned integers. My rule of thumb is:

* Use int for things that are unlikely to ever be big enough to worry about an int32 overflow (like the number of columns in a table -- if you had 2 billion columns, you would probably run into other problems sooner)
* Use `int64_t` for things that are expected to be able to overflow a 32-bit signed integer
* Use `size_t` exclusively when dealing with malloc, memcpy, and other C standard library functions, and for handling C++ loop indices in the STL (`for (size_t i = 0; i < foo.size(); ++i)`)

In Google's C++ style guide they say

```
Some people, including some textbook authors, recommend using unsigned types to represent numbers that are never negative. This is intended as a form of self-documentation. However, in C, the advantages of such documentation are outweighed by the real bugs it can introduce. Consider:

for (unsigned int i = foo.Length()-1; i >= 0; --i) ...

This code will never terminate! Sometimes gcc will notice this bug and warn you, but often it will not. Equally bad bugs can occur when comparing signed and unsigned variables. Basically, C's type-promotion scheme causes unsigned types to behave differently than one might expect.

So, document that a variable is non-negative using assertions. Don't use an unsigned type.
```

**Comment by gfyoung** [comment]
> Very honestly, I'm not sure I'm comfortable with this many unsigned integers.

That's fair.  In that case, `ssize_t` would be preferable to `size_t`.

**Comment by wesm** [comment]
`ssize_t` isn't a standard integer type; it's a POSIX typedef. For best cross-platform compatibility the best integer to use for values that are expected to be large is `int64_t`

File: pandas/_libs/parsers.pyx (hunk 1 above)
Comment (jeffknupp) [inline pandas/_libs/parsers.pyx]: If I could do that and keep in under 80 columns, I would have

File: pandas/_libs/parsers.pyx (hunk 3 above)
Comment (jeffknupp) [inline pandas/_libs/parsers.pyx]: In some specific places, `-1` is used as a sentinel value. This patch _should_ exhaustively identify all of those cases or change them to not need to use a negative number.

File: pandas/_libs/parsers.pyx (hunk 1 above)
Comment (gfyoung) [inline pandas/_libs/parsers.pyx]: Why can't you move the "pointer to data to be processed" one further right?

File: pandas/_libs/parsers.pyx (hunk 3 above)
Comment (gfyoung) [inline pandas/_libs/parsers.pyx]: Ah, I see.  Also, I think in light of what @wesm said <a href="https://github.com/pandas-dev/pandas/pull/17040#issuecomment-316812706">below</a>, it might be preferable to use `int64_t` instead given the cross-platform compatibility issue.

**Comment by jeffknupp** [comment]
@wesm FWIW, almost all of the changed fields are `malloc`-ed, `memcpy`-ed or `realloc`ed (the source of the original bug). I tried to keep the surface area of the change as small as possible but still address the places where a) an int32_t would overflow and b) one would reasonably expect and `(s)size_t`.

For me, `size_t` was always seemed natural because a) it's what `malloc()` expects and b) it's what `sizeof()` returns. The fact that it is unsigned is simply an implementation detail (albeit one that is always true).

That said, it sounds like you're -1 on size_t, which is fine. I'll fix all changed types to be `int64_t`.

File: pandas/_libs/parsers.pyx (hunk 1 above)
Comment (jeffknupp) [inline pandas/_libs/parsers.pyx]: Oops! Sorry, was looking at wrong file! Will fix in next push.

**Comment by jeffknupp** [comment]
As an aside, Google's style guide seems a bit all-over-the-map. They suggest not using unsigned types and in the very same breath dictate that assertions should be preferred. Of course, the only problem with that is `assert` is a macro dependent on the `NDEBUG` macro which itself is not part of C++ standard library. Also, I can't imagine finding a widely used version of `gcc` (well, `g++` I guess) that wouldn't warn about the type difference in a comparison of the form they present. If the compiler can't make *that* determination, you may as well just give up... :trollface: 

Enough bike-shedding, though! I made the requested changes and pushed. Please review at your convenience.

File: pandas/_libs/parsers.pyx [hunk 4]
```diff
@@ -121,30 +121,30 @@ cdef extern from "parser/tokenizer.h":
...
-        int *word_starts # where we are in the stream
-        int words_len
-        int words_cap
+        int64_t *word_starts # where we are in the stream
+        int64_t words_len
+        int64_t words_cap
 
-        char *pword_start    # pointer to stream start of current field
-        int word_start       # position start of current field
+        char *pword_start       # pointer to stream start of current field
```
Comment (gfyoung) [inline pandas/_libs/parsers.pyx]: So close!  One more space.

File: pandas/_libs/parsers.pyx [hunk 5]
```diff
@@ -178,13 +178,13 @@ cdef extern from "parser/tokenizer.h":
         char thousands
 
         int header # Boolean: 1: has header, 0: no header
-        int header_start # header row start
-        int header_end # header row end
+        int64_t header_start # header row start
+        int64_t header_end # header row end
```
Comment (gfyoung) [inline pandas/_libs/parsers.pyx]: I'm going to let OCD take over and suggest that you add a couple more spaces to align with the hashtag above. :smile:

**Comment by gfyoung** [comment]
@jeffknupp : Do you have a test to reproduce the issue (I know that was problematic was the issue, but I'm curious what the current status on that is now)?

Also, add an entry in the `whatsnew` .

**Comment by gfyoung** [comment]
@jeffknupp : Can you check that the issue in #14696 is handled by your changes?

**Comment by jeffknupp** [comment]
Yes, it is handled by this change.

Thanks,
Jeff

**Comment by wesm** [comment]
@jeffknupp sorry about _my_ bike shedding. I agree it's often a lot of worrying over nothing but thanks for indulging my OCDs

**Comment by jeffknupp** [comment]
@gfyoung Added what I think is an appropriate set of entries in `whatsnew` as well as some heavy-duty horizontal alignment 😉.

I have a very simple test (which I gave to @wesm and believe was linked to from the original issue) which would cause a simple script of the form:

```
import pandas as pd

try:
    df = pd.read_csv('test_file.gz2', low_memory=False)
except:
    print('Failed')
```

to core (and, after the allocation issue was fixed, to simply print an "out of memory" error). It required a data file that was ~300 MB after bzip2, for which there should be a link.

**Comment by gfyoung** [comment]
> as well as some heavy-duty horizontal alignment

It's beautiful, thanks. :smile: 

> I have a very simple test

Ah, okay.  It seems like we would have to add this to the repository, though I don't want to (nor do  commit something like that given the file size.  Thoughts?

**Comment by jeffknupp** [comment]
@wesm was faced with a similar issue for a similar problem in apache/arrow, which is where this first manifest itself. Let me dig around and see what he did (though I remember an email in the past few days specifically mentioning that the test included would only be run if a specific flag was set, and wouldn't run by default).

**Comment by jeffknupp** [comment]
So, it looks like it works aside from Python 2.7 on Windows? I'm looking at the failing tests now in appveyor. Would like to get this buttoned up tonight but may have to punt to the morning.

File: pandas/_libs/parsers.pyx [hunk 6]
```diff
@@ -121,30 +121,30 @@ cdef extern from "parser/tokenizer.h":
         io_callback cb_io
         io_cleanup cb_cleanup
 
-        int chunksize  # Number of bytes to prepare for each chunk
-        char *data     # pointer to data to be processed
-        int datalen    # amount of data available
-        int datapos
+        int64_t chunksize  # Number of bytes to prepare for each chunk
+        char *data        # pointer to data to be processed
```
Comment (jreback) [inline pandas/_libs/parsers.pyx]: pls line up

File: pandas/_libs/parsers.pyx [hunk 7]
```diff
@@ -1336,6 +1339,7 @@ cdef class TextReader:
         kh_destroy_str(table)
 
     cdef _get_column_name(self, Py_ssize_t i, Py_ssize_t nused):
+        cdef int64_t j
```
Comment (jreback) [inline pandas/_libs/parsers.pyx]: was this not defined before?

File: pandas/_libs/src/parser/tokenizer.c [hunk 8]
```diff
@@ -507,7 +517,7 @@ static int end_line(parser_t *self) {
             TRACE((
                 "end_line: ERROR!!! self->lines(%zu) >= self->lines_cap(%zu)\n",
                 self->lines, self->lines_cap))
-            int bufsize = 100;
+            size_t bufsize = 100;
             self->error_msg = (char *)malloc(bufsize);
             snprintf(self->error_msg, bufsize,
```
Comment (jreback) [inline pandas/_libs/src/parser/tokenizer.c]: I thought we were going to pre-allocate the error msg buffer?

File: pandas/_libs/src/parser/tokenizer.c [hunk 9]
```diff
@@ -1281,20 +1299,26 @@ int parser_trim_buffers(parser_t *self) {
     }
 
     /* trim line_start, line_fields */
-    new_cap = _next_pow2(self->lines) + 1;
-    if ((int)new_cap < self->lines_cap) {
+    if (new_cap < 1024 * 1024 * 1024) {
```
Comment (jreback) [inline pandas/_libs/src/parser/tokenizer.c]: why are you doing this ?

**Review by jreback — CHANGES_REQUESTED** [review]
so if you can add a test which triggers this before the fixes and works after. ideally you simply generate in memory a large enough string that you need. (and mark this as a slow test). There are a couple of examples where we do something this already in the parser tests.

File: pandas/_libs/src/parser/tokenizer.c (hunk 9 above)
Comment (jeffknupp) [inline pandas/_libs/src/parser/tokenizer.c]: Sorry, that must have been left over from old test code; will revert.

File: pandas/_libs/src/parser/tokenizer.c (hunk 8 above)
Comment (jeffknupp) [inline pandas/_libs/src/parser/tokenizer.c]: I think doing so is a logically separate change and has some subtleties (e.g. what should allocated size be and what happens when more is needed down the road but it's not obvious someone has to check their (possibly dynamic) output length against the buffer size). Feel free to open an issue and I'll submit a patch if I have time.

File: pandas/_libs/parsers.pyx (hunk 7 above)
Comment (jeffknupp) [inline pandas/_libs/parsers.pyx]: No, and for the life of me I couldn't figure out how it was working given all other instances seem to require it (left as an exercise for the reader as to why it worked but should be defined with the proper type here, regardless).

**Comment by jeffknupp** [comment]
@jreback Right, that's the testing solution I originally proposed on the issue. Marking it as a "slow" test, though, won't work, as someone looking to be exhaustive in their testing who includes the slow tests will be surprised when one fails because "slow" actually meant "bit" in this case. 

Is there already a separate category for "large memory" tests? It looks like that's what @wesm did for a very similar issue involving the same test data in apache/arrow

**Comment by jeffknupp** [comment]
Also, I have a reasonable idea of the fixes required to get the appveyor tests to pass but is there any way of testing these fixes without direct access to a windows environment? I don't want to get into a multi-hour push change, wait for appveyor to test PR, apply small fix, push change dev cycle here...

**Comment by gfyoung** [comment]
> Also, I have a reasonable idea of the fixes required to get the appveyor tests to pass but is there any way of testing these fixes without direct access to a windows environment?

Spinning up a Windows VM might be your way to go.  Then you can run the tests directly in a Windows environment on your computer.

However, I think running tests on `Appveyor` is the only way to go ATM if you don't want to go through all that work.  What you could do (to avoid having to wait for other builds) is to make an account on AppVeyor and make a new branch that mirrors this one in the PR.  Then you you can make whatever changes you want on that branch to get tests to pass (without competing with other PR's) and then merge them into the branch for this PR.

**Comment by gfyoung** [comment]
As I write this, I notice that AppVeyor is passing for you now, but Travis is broken, so I guess the issue is kind of moot for now. :smile:

**Comment by jeffknupp** [comment]
@wesm @jreback is it possible to merge this and leave the original issue open until I submit a suitable patch to demonstrate a negative test for this change? Basically, I need to get this merged into pandas ASAP so that a) I can do a dev install from the official pandas repo in my project and b) want to stop doing a) ASAP, and I don't want the actual fix to fall by the wayside as we determine the best way to test.

**Comment by jeffknupp** [comment]
(Or merge and leave the PR open with the understanding I will work on the test this weekend or early next week)

**Comment by gfyoung** [comment]
@jeffknupp : The next release (`0.21`) won't come out until end of September AFAIK, so I don't think there is any rush to merge this.  Feel free to take your time on this :smile:

**Comment by gfyoung** [comment]
I've tagged it for `0.21` so we won't forget it :smile:

**Comment by jreback** [comment]
you can't simply install from your fork the same way you would from master just a different branch

File: pandas/_libs/src/parser/tokenizer.c (hunk 8 above)
Comment (jreback) [inline pandas/_libs/src/parser/tokenizer.c]: well if you like to open an issue that is fine, the original PR was closed and didn't have an issue.

File: pandas/_libs/src/parser/tokenizer.c [hunk 10]
```diff
@@ -712,15 +718,17 @@ int skip_this_line(parser_t *self, int64_t rownum) {
...
+int tokenize_bytes(parser_t *self, size_t line_limit, int64_t start_lines) {
+    int64_t i, slen;
     int should_skip;
     char c;
     char *stream;
     char *buf = self->data + self->datapos;
 
     if (make_stream_space(self, self->datalen - self->datapos) < 0) {
-        self->error_msg = "out of memory";
+        size_t bufsize = 100;
```
Comment (jreback) [inline pandas/_libs/src/parser/tokenizer.c]: so was this pre-allocated before? its seems *some* (but not all ) errors message buffers were pre-allocated

File: pandas/_libs/src/parser/tokenizer.c [hunk 11]
```diff
@@ -1227,34 +1235,40 @@ int parser_trim_buffers(parser_t *self) {
...
         } else {
-            self->word_starts = (int *)newptr;
+            self->word_starts = (int64_t *)newptr;
             self->words_cap = new_cap;
         }
     }
 
     /* trim stream */
-    new_cap = _next_pow2(self->stream_len) + 1;
+    if (new_cap < INT32_MAX) {
```
Comment (jreback) [inline pandas/_libs/src/parser/tokenizer.c]: what's the rationale here?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
note that we DO test on linux-32, but on nightly (and no way to actually test this a-priori, so would only show up after merging).

e.g. see https://travis-ci.org/MacPython/pandas-wheels

File: pandas/_libs/src/parser/tokenizer.c (hunk 11 above)
Comment (jeffknupp) [inline pandas/_libs/src/parser/tokenizer.c]: Sorry, you found the last remaining holdout of early changes I made while determining what was causing the "oom" issue. This code is especially non-sensical since both branches do the same thing, but it's just cruft that needs to be removed at this point (and is in the latest commit).

File: pandas/_libs/src/parser/tokenizer.c (hunk 10 above)
Comment (jeffknupp) [inline pandas/_libs/src/parser/tokenizer.c]: `error_msg` is _not_ pre-allocated (in the "allocated when the parser object is created sense"). It appears one is _meant_ to allocate memory for the error message just before setting it, which of course is error prone and led to the double-free issue.

Like I said, for now, I've just made sure that all uses of `error_msg` are directly preceeded by an allocation (I think two of about ten were missing them). A better approach would be to allocate a reasonable amount of space when the parser is created and not worry about allocation anywhere else, but that should be in its own PR (which I'm working on ATM).

**Comment by jeffknupp** [comment]
@jreback Note that the issue is not related to the underlying architecture of the machine. Rather, it's due to the fact that using `int`s in calls to `malloc`/`realloc` exposes you to integer overflow issues when trying to allocate anything more than `INT_MAX / 2` bytes (a 32-bit system can, by definition, store an address greater than 2^16).

To trigger the issue, one simply needs to create a CSV where at least one column is larger than `INT_MAX` bytes. However, since that value is typically exactly 2 GB, you'll need a machine with at least that much RAM. In reality, you'll need quite a bit more since a) you need to make at least one copy of the data for the parser and b) you need to make sure you don't start swapping so bad you take down the system while also avoid being killed by the `oom killer`.

So, tl;dr, any test for this issue will need to be run on a machine with sufficient RAM, where "sufficient" is likely in the > 6GB range.

**Comment by jeffknupp** [comment]
@jreback @gfyoung OK, this should be considered complete.

I added a new pytest option, `--run-highmemory` which _enables_ tests marked `high_memory`. They are not run by default as I don't want to destroy your numerous CI environments 😄 .

The new test, marked as `high_memory`, is crude but effective. The whole test takes roughly 8GB on my MBP and maybe 20 seconds to complete. I have confirmed:

- If run against current `master`, the `pytest` crashes (which I think we can consider "a failure") as was observed in the issue
- If run on the new code with `--run-highmemory` enabled, the test passes

Note that since I'm not aware of a nontrivial way of constructing a test to see if the interpreter itself will crash, test coverage will likely complain that the two new `error_msg` allocations are not covered (since this code fixes both the crashes at those locations and the code that was incorrectly invoking them in the first place). I hope that's OK as I'm not sure how much more time I can devote to this specific issue.

**Comment by gfyoung** [comment]
> I hope that's OK as I'm not sure how much more time I can devote to this specific issue.

Understood.  Could you at least patch the lint errors on Travis right now?  It would be great to get all builds passing. :smile:

**Comment by jeffknupp** [comment]
@gfyoung because you asked so nicely... 😄 . A bit of acrobatics for 80-character width. `120` is the new `80`.

File: pandas/tests/io/parser/test_parsers.py [hunk 12]
```diff
@@ -25,6 +28,18 @@
 from .dtypes import DtypeTests
 
 
+@pytest.mark.high_memory
+def test_bytes_exceed_2gb():
```
Comment (gfyoung) [inline pandas/tests/io/parser/test_parsers.py]: I would refactor to add this into the `CParser` tests.  You don't necessarily have to have the test run for the low-memory c-parser (you can just have it skip if it is the case).

File: pandas/tests/io/parser/test_parsers.py [hunk 13]
```diff
@@ -25,6 +28,18 @@
...
+    """Read from a "CSV" that has a column larger than 2GB.
+
+    GH 16798
+    """
+    csv = StringIO('strings\n' + '\n'.join(
+        ['x' * (1 << 20) for _ in range(2100)]))
+    df = read_csv(csv, low_memory=False)
+    assert not df.empty
```
Comment (gfyoung) [inline pandas/tests/io/parser/test_parsers.py]: Ideally, it would be great if we could check the actual contents.  Then again, I'm not sure I want to instantiate another copy of this `DataFrame` in memory 😄

File: doc/source/whatsnew/v0.21.0.txt [hunk 14]
```diff
@@ -201,7 +201,8 @@ I/O
 ^^^
 
 - Bug in :func:`read_csv` in which non integer values for the header argument generated an unhelpful / unrelated error message (:issue:`16338`)
-
+- Bug in :func:`read_csv` in which passing a CSV with at least one very large (i.e. more than 2^31 - 1 bytes) column along with ``low_memory=False`` would cause an integer overflow. The result was an always unsuccessful attempt to allocate an enourmous buffer and then reporting "Out of memory." (:issue:`16798`).
+- Bug in :func:`read_csv` in which some errors paths were assigning error messages to the internal tokenizer's ``error_msg`` field without first allocating the memory. When this happened as part of exception handling, it resulted in a double ``free`` and the program halted due to a ``SIGSEGV`` (:issue:`16798`).
```
Comment (gfyoung) [inline doc/source/whatsnew/v0.21.0.txt]: I wonder if we can make these descriptions less technical.  Our (less technically) audience might not understand everything written here.  For example, you could write:
~~~
- Bug in :func:`read_csv` in which passing a CSV with at least one very large...would cause Python to run out of memory.
- Bug in :func:`read_csv` in which memory mismanagement with error messages caused Python to crash
~~~

True, it doesn't capture the entire error description as before, but it gets the general idea across I hope.

File: pandas/tests/io/parser/test_parsers.py (hunk 13 above)
Comment (jreback) [inline pandas/tests/io/parser/test_parsers.py]: this is fine - we care about segfault

File: doc/source/whatsnew/v0.21.0.txt (hunk 14 above)
Comment (jeffknupp) [inline doc/source/whatsnew/v0.21.0.txt]: Sure. I was a bit hesitant to write these entries for this exact reason (I wasn't sure who the audience was and what the expectations were about entries). I've removed a lot of the detail while still giving the reader enough information to determine if they bug they just saw is, in fact, one of the two listed.

File: pandas/tests/io/parser/test_parsers.py (hunk 12 above)
Comment (jeffknupp) [inline pandas/tests/io/parser/test_parsers.py]: I considered that, but given that the parser tests in general are all constructed to run all of the tests under one of three parser setups (and this only applies to one of them), it seemed like it would be shoe-horning it in just to keep things "pretty". There were a few other locations where one-off tests for the parsers were created because they had very specific functions, so I thought this was closer to the latter case.

There was also text in the documentation that pointed to the old `unittest` style tests being deprecated in favor of normal `pytest`-style tests.

Of course, all that said, I'm happy to change if you disagree.

File: pandas/tests/io/parser/test_parsers.py (hunk 13 above)
Comment (jeffknupp) [inline pandas/tests/io/parser/test_parsers.py]: Thanks. Would this be mergeable, then? I've addressed @gfyoung 's comments which I believe were the last outstanding.

**Review by jreback — APPROVED** [review]
lgtm merge in green

**Comment by jeffknupp** [comment]
Finally all green. Thanks for your help @jreback @gfyoung @wesm @codecov[bot] @pep8speaks ...

**Comment by jreback** [comment]
thanks @jeffknupp

**Comment by wesm** [comment]
thanks @jeffknupp!!

**Comment by jreback** [comment]
ok this is causing the 32-bit builds to segfault

https://travis-ci.org/MacPython/pandas-wheels/jobs/256710160

no easy way to debug this in travis

can u setup a vm and see what is happening ?

thanks

**Comment by gfyoung** [comment]
@jreback : We could always just drop support for 32-bit.  Then this won't be an issue. 😄 

On a more serious note, it would be great if we could somehow add 32-bit testing to Travis.  I think `numpy` might have a way to do it without using Docker (the current workaround based on what I've seen on Travis issues), but I had issues previously setting it up.  Perhaps it might be worthwhile to revisit.

In any case, I suspect the issue might be the `malloc`-ing for error messages (note that it uses `size_t` instead of `int64_t` unlike the other changes).  I don't see at this point how the other changes could impact this.  What do you think?

**Comment by jeffknupp** [comment]
I'm not sure how/why a 32-bit Travis worker would behave any differently than a 64-bit one in this case. Unless the 32-bit worker had less RAM, there's no reason why it should segfault (unless there is a regression somewhere else?).

Also, @gfyoung , just FYI, `malloc(3)` is _meant_ to take a `size_t` (for obvious reasons) as seen in its declaration (and those of all the other memory allocation related functions):

```
SYNOPSIS
     #include <stdlib.h>

     void *
     calloc(size_t count, size_t size);

     void
     free(void *ptr);

     void *
     malloc(size_t size);

     void *
     realloc(void *ptr, size_t size);

     void *
     reallocf(void *ptr, size_t size);

     void *
     valloc(size_t size);
```

**Comment by jeffknupp** [comment]
@jreback Also, AFAICT, you weren't actually running the new test, since `--run-high-memory` wasn't passed as an argument to `py.test` according to the Travis run logs...

**Comment by jreback** [comment]
I created #17063 to track this.

@jeffknupp 

you new test is not enabled, but this is not the issue. a previously passing 32-bit build now segfaults. Pls try to replicate / fix this issue. If this cannot be tracked down soon I will revert this commit.

We currently DO support 32-bit builds. That may be re-evaluateda at some point (even for 0.21.0). But that should not couple with this issue.

**Comment by gfyoung** [comment]
> Also, @gfyoung , just FYI, malloc(3) is meant to take a size_t (for obvious reasons)

@jeffknupp : Doh!  You're absolutely right.  Then again, segfaulting with this commit is defying logic as it is right now. 😄

**Comment by jeffknupp** [comment]
@jreback @gfyoung is there any way for us to discuss this in more detail via a more synchronous form of communication? I'm a bit confused about the failure @jreback pointed to and how one can match up what ran in each MacPython/pandas-wheels build.

Do you guys have Slack/IRC/AIM/ICQ/Prodigy that we could sync up on?

**Comment by gfyoung** [comment]
> is there any way for us to discuss this in more detail via a more synchronous form of communication?

To have more synchronous communication, you need to be all available at the same time :smile: .  Unfortunately, that's possible ATM.

> Do you guys have Slack/IRC/AIM/ICQ/Prodigy that we could sync up on?

Not that I know of.  We have Gitter, though I'm not sure that's going to alleviate the issue.

> I'm a bit confused about the failure @jreback pointed to and how one can match up what ran in each MacPython/pandas-wheels build.

You're not alone in the confusion. :cry: I don't think any of us have a clear of what the failure is, as debugging on Travis is very difficult.  What @jreback is suggesting is that you spin up a 32-bit Linux VM and try running tests in that VM to see if you can reproduce the segfault when running tests.
